### ISSUE TITLE
Grid: getDataProvider() inside setItems() returns other instance

### ISSUE THREAD

**Issue by MathiasAugstein** [vaadin/flow#12384]
I just stumbled upon Grids' DataProviderListener-feature and wanted to use it, and unfortunately it looks like I found a bug. 
### Description of the bug
When using `getDataProvider()` inside overridden `setItems()`, one has to be sure to call `getDataProvider()` after `super.setItems()`, as setting new items creates a new DataProvider. But in the MWE below it shows that `getDataProvider()` returns a different instance. According to the debugger it is neither the old nor the new instance.
### Minimal reproducible example
```
public class MainLayout extends VerticalLayout
{
	public MainLayout()
	{
		Grid<Integer> grid = new CustomGrid<>();
		grid.addColumn(i -> i.toString());
		List<Integer> list = IntStream.range(0, 10).boxed().collect(Collectors.toList());
		grid.setItems(list);
		add(grid);
                // should trigger the listener, but does not
		list.add(10);
	}
	
	public class CustomGrid<T> extends Grid<T> {

		@Override
		public void setItems(Collection<T> items) {
			super.setItems(items);
			getDataProvider().addDataProviderListener(ev -> System.out.println("Triggered"));
		}
		
	}
}
```
### Expected behavior
"Triggered" appears on the command line.
### Actual behavior
Nothing happens.
### Versions:
    - Vaadin version: 14.7.5
    - Java version: 1.8
    - OS version: Win10

**Comment by caalador** [comment]
There is no code that is looking at the Collection if it is manipulated or not as there is no such feature for Collections.
When manipulating the list you should call `grid.getDataProvider().refreshAll();` and with this the added listener will be called.

**Comment by MathiasAugstein** [comment]
Good point, adding your line does trigger the DataProviderListener. Seems like either there is something missing in the MWE that is needed to trigger the bug, or I simply had another error in my code that I did not see (and therefor assume there is a bug).

The problem is, I did not commit my original code (as it was not working) and therefor cannot reproduce the problem right now. I will close this issue and keep an eye open the next time I use `getDataProvider` inside `setItems`. Sorry for bothering you.
